### PR TITLE
Upgrades from yarn 1.x to yarn berry (4.x)

### LINKED ISSUE #144: upgrade to yarn berry
yarn classic (1.x) is deprecated move to yarn berry (4.0)

### PR DESCRIPTION
**Pull request by arthurgousset**
### Description

Upgrades our yarn package manager from 1.x to "yarn berry" (current yarn 4.x).

```[tasklist]
### [Migration steps](https://yarnpkg.com/migration/guide)
- [x] Make sure you're using Node 18+
- [x] Run `corepack enable` to activate Corepack
- [x] Go into your project directory
- [x] Run `yarn set version berry`
- [x] Convert your `.npmrc` and `.yarnrc` files into `.yarnrc.yml`
- [x] Run `yarn install` to migrate the lockfile
- [x] Commit all changes
```

### References

- https://yarnpkg.com/migration/guide
- https://github.com/yarnpkg/berry/issues/4908
- https://github.com/celo-org/social-connect/pull/73
- https://github.com/yarnpkg/berry/issues/821

### Tested

_An explanation of how the changes were tested or an explanation as to why they don't need to be._

```[tasklist]
### Top level `./package.json`
- [x] "dev" script - `yarn run dev` works as expected, UI can be seen on local host
- [x] "lint" script - `yarn run lint` works as expected "✔ No ESLint warnings or errors "
- [x] ~~"setup" script~~ - Deleted (see justification in comments)
- [x] "prepare" script - works as expected, `husky` git hooks are installed. This runs with `yarn install`, but I also tested it separately
- [x] "prettify" script - works as expected
``` 

```[tasklist]
### Web workspace `web/package.json`
- [x] "setup" script - Deleted (see justification in comments)
- [x] "dev" script - `yarn run dev` works as expected, UI can be seen on local host
- [x] "build" script - `yarn run build` works as expected, output in `web/.next/`
- [x] "start" script - `yarn run start` works as expected, UI can be seen on local host
- [x] "lint" script - `yarn run lint` works as expected, "✔ No ESLint warnings or errors"
```

```[tasklist]
### Firebase workspace `firebase/package.json`
- [x] "preserve" script - `yarn run preserve` works as expected, builds the TS project in `/dist`
- [x] "build" script - `yarn run build` works as expected, builds the TS project in `/dist`
- [x] "clean" script - `yarn run clean` works as expected, contents of `/dist` are deleted
- [x] "lint" script - `yarn run lint` works as expecrted, tested by adding an unused variable.
- [x] "serve" script - `yarn run serve` works as expected (needed to `firebase login` first, which I documented in the README). Not clear how to test the server works as expected, but outside of scope in this PR, left a note in comments below.
- [x] "build:rules" script - `yarn run build:rules` works as expected, creates `database-rules.json` file
- [x] "transfer-funds" script - `yarn run transfer-funds` runs, but exits with a TS compilation error. Already happening on `master` (yarn 1.x) so not related to yarn berry upgrade. Not fixing in this PR, but left a note in comments below.
- [x] "cli" script - `yarn run cli` works as expected, tested `yarn run cli accounts:get --net alfajores` and `yarn cli config:get` for example.
- [x] "keys:encrypt" script - `yarn run keys:encrypt` runs as expected (for the purposes of yarn berry upgrade), but stops because I don't have `gcloud` installed. Not sure we're still managing secrets like that, left a comment below.
- [x] "deploy:staging" script - `yarn run deploy:staging` runs as expected. I don't have sufficient Firebase permissions, but for the purposes of yarn berry upgrade, this is sufficient.
- [x] "deploy:prod" script - didn't test to avoid deploying to production, but `firebase anyCommand` work as expected with yarn berry upgrade.
``` 

### Related issues

- Fixes #144

### Backwards compatibility

This change is backwards compatible. All scripts work as expected or are updated accordingly.

### Documentation

No additional docs for yarn berry needed. The version is specified in the top level `package.json` under 

```json
"packageManager": "yarn@4.0.2"
```

### CODE REVIEW

**Comment by arthurgousset** [comment]
(changelog)

The "setup" script in `./package.json` that installs `vercel` globally is broken: `yarn global add vercel@latest`.

That's because [`yarn global` is deprecated in Yarn 2.x](https://yarnpkg.com/migration/guide#use-yarn-dlx-instead-of-yarn-global), which seems to be annoying quite a few people:
- https://github.com/yarnpkg/berry/issues/821

Looks like `vercel` is only used in `"setup": "vercel env pull"` in [`web/package.json`](https://github.com/celo-org/faucet/blob/ef952e8efbe8ec856966b6d703019e1e021a7bc6/apps/web/package.json#L6).

Potential solutions:
1. remove the "setup" script in `./package.json` and add `vercel` in `devDependencies`
2. remove the "setup" script in `./package.json` and modify `vercel env pull` in `web/package.json` to use `yarn dlx`

I chose option (2) to keep the number of actual dependencies low. `vercel` is only used to fetch `.env` variables once in case a dev wants to run the web app locally.

**Comment by arthurgousset** [comment]
Notes-to-self: 

1. Existing bug on `master`: `yarn run serve`: not clear how to test the local server works as expected

1. Existing bug on `master`: `yarn run transfer-funds`: exits with TS compilation error

  ```sh
  /Users/arthur/Documents/celo-org/faucet/apps/firebase/node_modules/ts-node/src/index.ts:859
      return new TSError(diagnosticText, diagnosticCodes, diagnostics);
             ^
  TSError: ⨯ Unable to compile TypeScript:
  scripts/transfer-funds.ts:91:44 - error TS2339: Property 'sendAndWaitForReceipt' does not exist on type 'false | CeloTransactionObject<boolean>'.
    Property 'sendAndWaitForReceipt' does not exist on type 'false'.
  
  91           console.log('receipt', await tx?.sendAndWaitForReceipt())
                                                ~~~~~~~~~~~~~~~~~~~~~
  ```
  
1. Existing bug on `master`: ESLint error that can't find `faucet/tsconfig.json` in multiple files

`Parsing error: Cannot read file '/users/arthur/documents/celo-org/faucet/tsconfig.json'.eslint`

**Comment by arthurgousset** [comment]
Testing that `firebase anyCommand` works in the `firebase/` workspace:

```diff
"scripts": {
+    "firebase-yarn-berry-test": "firebase --version"
  },
```

```
$ yarn run firebase-yarn-berry-test
13.0.3
```

**Comment by arthurgousset** [comment]
My last commit that merges master into my branch seems to have broken the build :(

**Comment by arthurgousset** [comment]
Maybe it has to do with this PR that was merged 6h ago:
- https://github.com/celo-org/faucet/pull/136

**Comment by arthurgousset** [comment]
`master` branch has a recent commit that downgrades from "@types/react": "18.2.48" to "@types/react": "18.2.3". 

I hypothesise that the build error originates here. 

Upgrading back to `"@types/react": "18.2.48"` fixed the build error:

```sh
$ cd apps/web
$ yarn add @types/react@18.2.48
$ yarn run build     
✓ Linting and checking validity of types    
 ✓ Creating an optimized production build    
 ✓ Compiled successfully
 ✓ Collecting page data    
 ✓ Generating static pages (2/2) 
 ✓ Collecting build traces    
 ✓ Finalizing page optimization    

Route (pages)                              Size     First Load JS
┌   /_app                                  0 B            93.8 kB
├ λ /[chain]                               14.6 kB         108 kB
├   └ css/ab4939cb46ff182c.css             2 kB
├ ○ /404                                   178 B            94 kB
├ λ /api/auth/[...nextauth]                0 B            93.8 kB
└ λ /api/faucet                            0 B            93.8 kB
+ First Load JS shared by all              94.5 kB
  ├ chunks/framework-8e279965036b6169.js   45.4 kB
  ├ chunks/main-3099200c5c7ddf87.js        32.9 kB
  ├ chunks/pages/_app-331eabd72a9d2c3b.js  13.8 kB
  ├ chunks/webpack-266e854748dabb79.js     1.74 kB
  └ css/442e2a1785394bda.css               669 B

λ  (Server)  server-side renders at runtime (uses getInitialProps or getServerSideProps)
○  (Static)  automatically rendered as static HTML (uses no initial props)
```

File: package.json [hunk 1]
```diff
@@ -8,13 +8,11 @@
...
   "workspaces": [
-    "apps/*",
-    "packages/*"
+    "apps/*"
   ],
   "scripts": {
     "dev": "yarn --cwd apps/web dev",
     "lint": "yarn --cwd apps/web lint --fix && yarn --cwd apps/firebase lint --fix",
-    "setup": "yarn global add vercel@latest",
     "prepare": "husky install",
```
[the remark this answers is not in the document]
Comment (aaronmgdr) [reply package.json:16]: yarn 2 meaning 2 or greater

[the remark this answers is not in the document]
Comment (arthurgousset) [reply package.json:16]: Great to know, I'll add that 👍

[the remark this answers is not in the document]
Comment (arthurgousset) [reply package.json]: That's done in https://github.com/celo-org/faucet/pull/147/commits/67d4266a99a359ead8507a60c13d017e5877d391

**Comment by arthurgousset** [comment]
I noticed `husky` is a dev dependency in:

- the top level `package.json`, and separately

  ```json
  "devDependencies": {
  
      "husky": "^8.0.0",
  
    },
  ```

- in the `firebase/` workspace `package.json`

    ```json
    "devDependencies": {
      
        "husky": "^8.0.3",
       
      },
    ```

I'm wondering if I can delete `husky` from the dev dependencies in the `firebase/` workspace.
I don't see any husky configs in the `firebase` repo.

**Comment by arthurgousset** [comment]
Also asked the socket team about an alert that is inconsistent on [Slack](https://celo-org.slack.com/archives/C04019VGSDA/p1705944675080499):

> I'm seeing a [socket.dev](http://socket.dev/) alert in the following PR: [Upgrades from yarn 1.x to yarn berry (4.x) #147](https://github.com/celo-org/faucet/pull/147/checks?check_run_id=20738033571)
> It states that [npm/archiver-utils@2.1.0](https://socket.dev/npm/package/archiver-utils/overview/2.1.0) has an invalid package.json (see screenshot).
> When I follow the [hyperlink](https://socket.dev/npm/package/archiver-utils/issues/2.1.0?issue=invalidPackageJSON) or check any issues in [npm/archiver-utils@2.1.0](https://socket.dev/npm/package/archiver-utils/overview/2.1.0) (at [socket.dev/npm/package/archiver-utils/issues/2.1.0](http://socket.dev/npm/package/archiver-utils/issues/2.1.0)) I can't find any issues.
> Am I missing something? [Socket.dev](http://socket.dev/) flagged this issue on 3 separate reports I ran when committing to this PR.
> Thanks in advance :+1:
> 
> image
> 
> image

**Comment by arthurgousset** [comment]
Noting that I checked [these two socket alerts](https://github.com/celo-org/faucet/pull/147/checks?check_run_id=20738033571). The use of native code looks good to me.

image

**Comment by aaronmgdr** [comment]
agree no need for husky in firebase package. just in root

**Comment by arthurgousset** [comment]
Removed `husky` in `firebase/` workspace in https://github.com/celo-org/faucet/pull/147/commits/efc5822ae401469de21398528e7f9bac4e2275a9

**Comment by arthurgousset** [comment]
Now you can re-approve @aaronmgdr (or @celo-org/devtooling) 👍

**Pull request merged**